lib/levanter/src/levanter/layers/attention.py

```diff
@@ -1329,124 +1337,147 @@ def flatten(axes):
+
+    # K should not be sharded for splash attention
+    if physical_axes_k[2] is not None:
+        raise NotImplementedError(
+            "Splash attention does not support sharding the KV sequence dimension. "
+            f"Got KV sequence spec: {physical_axes_k[2]}"
+        )
+
+    # Compute block sizes based on per-shard sequence lengths
+    shard_Sq = max(1, Sq // max(1, q_seq_shards))
+    shard_Sk = max(1, Sk // max(1, kv_seq_shards))
+
+    def _compatible_block(shard_len: int, max_block: int) -> int:
+        """Pick largest block <= max_block that divides shard_len; prefer multiples of 128."""
+        if shard_len <= 0:
+            return max_block
+        cap = min(max_block, shard_len)
+        for step in (128, 1):
+            candidate = cap - (cap % step)
+            while candidate >= step:
+                if shard_len % candidate == 0:
+                    return candidate
+                candidate -= step
+        return 1
+
+    block_q = _compatible_block(shard_Sq, block_size)
+    block_kv = _compatible_block(shard_Sk, block_size)
+
+    block_sizes = splash_attention_kernel.BlockSizes(
+        block_q=block_q,
+        block_kv_compute=block_kv,
+        block_kv=block_kv,
+        block_q_dkv=block_q,
+        block_kv_dkv=block_kv,
+        block_kv_dkv_compute=block_q,
+        block_q_dq=block_q,
+        block_kv_dq=block_kv,
+    )
+
+    # Create mask with GLOBAL shapes (outside shard_map)
+    if mask is None:
+        base_mask = splash_attention_mask.FullMask(_shape=(Sq, Sk))
+    elif isinstance(mask, AttentionMask):
+        if mask.is_causal:
+            if mask.causal_offset is not None:
+                raise NotImplementedError(
+                    "Causal offsets are not supported for splash attention. Please use a standard causal mask."
+                )
+            base_mask = splash_attention_mask.CausalMask((Sq, Sk), offset=0, shard_count=q_seq_shards)
+        else:
+            base_mask = splash_attention_mask.FullMask(_shape=(Sq, Sk))
+        if mask.sliding_window is not None:
+            local_mask = splash_attention_mask.LocalMask(
+                shape=(Sq, Sk),
+                window_size=(mask.sliding_window - 1, None),
+                offset=0,
+                shard_count=q_seq_shards,
+            )
+            base_mask = splash_attention_mask.LogicalAnd(base_mask, local_mask)
+        if mask.explicit_mask is not None:
+            raise NotImplementedError("Explicit masks are not yet supported for splash attention")
+    elif isinstance(mask, NamedArray):
+        raise NotImplementedError("NamedArray masks are not yet supported for splash attention")
+    else:
+        raise ValueError(f"Unknown mask type: {mask}")
+
+    kernel_mask = splash_attention_mask.MultiHeadMask(masks=[base_mask for _ in range(Hq)])
+
+    # Create kernel with GLOBAL shapes and q_seq_shards (outside shard_map)
+    splash_kernel = splash_attention_kernel.make_splash_mha(
```

Comment:
This has pretty little to do with this PR, but is the kernel construction JITed across batches?

Reply:
i think so. jit will just examine the shapes of the inputs and the hash of the static pytree bits and assume nothing else changed